File: src/main/java/cn/academy/vanilla/meltdowner/skill/ElectronMissile.java
/**
 * Copyright (c) Lambda Innovation, 2013-2015
 * 本作品版权由Lambda Innovation所有。
 * http://www.li-dev.cn/
 *
 * This project is open-source, and it is distributed under
 * the terms of GNU General Public License. You can modify
 * and distribute freely as long as you follow the license.
 * 本项目是一个开源项目，且遵循GNU通用公共授权协议。
 * 在遵照该协议的情况下，您可以自由传播和修改。
 * http://www.gnu.org/licenses/gpl.html
 */
package cn.academy.vanilla.meltdowner.skill;

import static cn.lambdalib.util.generic.RandUtils.*;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import cn.academy.ability.api.Skill;
import cn.academy.ability.api.ctrl.ActionManager;
import cn.academy.ability.api.ctrl.Cooldown;
import cn.academy.ability.api.ctrl.SkillInstance;
import cn.academy.ability.api.ctrl.action.SkillSyncAction;
import cn.academy.core.client.ACRenderingHelper;
import cn.academy.vanilla.meltdowner.client.render.MdParticleFactory;
import cn.academy.vanilla.meltdowner.entity.EntityMdBall;
import cn.academy.vanilla.meltdowner.entity.EntityMdRaySmall;
import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.networkcall.RegNetworkCall;
import cn.lambdalib.networkcall.s11n.StorageOption.Data;
import cn.lambdalib.networkcall.s11n.StorageOption.Instance;
import cn.lambdalib.networkcall.s11n.StorageOption.RangedTarget;
import cn.lambdalib.util.generic.VecUtils;
import cn.lambdalib.util.mc.EntitySelectors;
import cn.lambdalib.util.mc.WorldUtils;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.Vec3;
import net.minecraft.world.World;

/**
 * @author WeAthFolD
 */
@Registrant
public class ElectronMissile extends Skill {
	
	public static final ElectronMissile instance = new ElectronMissile();
	
	static int MAX_HOLD = 5;

	private ElectronMissile() {
		super("electron_missile", 5);
	}

	@Override
	public SkillInstance createSkillInstance(EntityPlayer player) {
		return new SkillInstance().addChild(new EMAction());
	}
	
	public static class EMAction extends SkillSyncAction {
		
		LinkedList<EntityMdBall> active;
		
		int ticks;
		
		@Override
		public void onStart() {
			super.onStart();
			
			if(!isRemote) {
				active = new LinkedList();
			}
		}
		
		@Override
		public void onTick() {
			if(!cpData.perform(instance.getOverload(aData), instance.getConsumption(aData)) && !isRemote)
				ActionManager.abortAction(this);
			
			if(!isRemote) {
				if(ticks > instance.callFloatWithExp("time_limit", aData))
					ActionManager.abortAction(this);
				
				if(ticks % 10 == 0) {
					if(active.size() < MAX_HOLD) {
						EntityMdBall ball = new EntityMdBall(player);
						world.spawnEntityInWorld(ball);
						active.add(ball);
					}
				}
				if(ticks != 0 && ticks % 8 == 0) {
					List<Entity> list = WorldUtils.getEntities(player, instance.callFloatWithExp("range", aData), 
						EntitySelectors.and(EntitySelectors.excludeOf(player), EntitySelectors.living));
					if(!active.isEmpty() && !list.isEmpty() && cpData.perform(
						instance.callFloatWithExp("overload_attacked", aData), 
						instance.callFloatWithExp("consumption_attacked", aData))) {
						double min = Double.MAX_VALUE;
						Entity result = null;
						for(Entity e : list) {
							double dist = e.getDistanceToEntity(player);
							if(dist < min) {
								min = dist;
								result = e;
							}
						}
						
						// Find a ramdom ball and destroy it
						int index = 1 + nextInt(active.size());
						Iterator<EntityMdBall> iter = active.iterator();
						EntityMdBall ball = null;
						while(index --> 0)
							ball = iter.next();
						iter.remove();
						
						// client action
						spawnRay(player, world, 
							VecUtils.entityPos(ball), 
							VecUtils.add(VecUtils.entityPos(result), VecUtils.vec(0, result.getEyeHeight(), 0)));
						
						// server action
						result.hurtResistantTime = -1;
						MDDamageHelper.attack(result, player, instance.callFloatWithExp("damage", aData));
						aData.addSkillExp(instance, instance.getFloat("expincr"));
						ball.setDead();
					}
				}
			} else
				updateEffect();
			
			++ticks;
		}
		
		@Override
		public void onEnd() {
			setCooldown(instance, instance.getFunc("cooldown").callInteger(ticks));
		}
		
		@Override
		public void onFinalize() {
			if(!isRemote) {
				for(EntityMdBall ball : active) {
					ball.setDead();
				}
			}
		}
		
		// CLIENT
		@SideOnly(Side.CLIENT)
		void updateEffect() {
			int count = rangei(1, 3);
			while(count --> 0) {
				double r = ranged(0.5, 1);
				double theta = ranged(0, Math.PI * 2);
				double h = ranged(-1.2, 0);
				Vec3 pos = VecUtils.add(VecUtils.vec(player.posX, player.posY + ACRenderingHelper.getHeightFix(player), player.posZ), VecUtils.vec(r * Math.sin(theta), h, r * Math.cos(theta)));
				Vec3 vel = VecUtils.vec(ranged(-.02, .02), ranged(.01, .05), ranged(-.02, .02));
				world.spawnEntityInWorld(MdParticleFactory.INSTANCE.next(world, pos, vel));
			}
		}
		
	}
	
	@RegNetworkCall(side = Side.CLIENT)
	static void spawnRay(@RangedTarget(range = 15) EntityPlayer _player, @Instance World world, @Data Vec3 from, @Data Vec3 to) {
		jobSpawnRay(world, from, to);
	}
	
	@SideOnly(Side.CLIENT)
	static void jobSpawnRay(World world, Vec3 from, Vec3 to) {
		EntityMdRaySmall ray = new EntityMdRaySmall(world);
		ray.setFromTo(from, to);
		world.spawnEntityInWorld(ray);
	}
	
}

File: src/main/scala/cn/academy/ability/vanilla/meltdowner/skill/ElectronMissile.scala
package cn.academy.ability.vanilla.meltdowner.skill

import cn.academy.ability.Skill
import cn.academy.ability.context.{ClientContext, ClientRuntime, Context, RegClientContext}
import cn.academy.client.render.particle.MdParticleFactory
import cn.academy.client.render.util.ACRenderingHelper
import cn.academy.entity.{EntityMdBall, EntityMdRaySmall}
import cn.lambdalib2.s11n.network.NetworkMessage.Listener
import cn.lambdalib2.util.{MathUtils, VecUtils}
import cn.lambdalib2.util.MathUtils._
import cn.lambdalib2.util.{EntitySelectors, WorldUtils}
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.math.Vec3d

/**
  * @author WeAthFolD, KSkun
  */
object ElectronMissile extends Skill("electron_missile", 5) {

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyID: Int) = activateSingleKey(rt, keyID, p => new EMContext(p))
  
}

object EMContext {

  final val MSG_EFFECT_SPAWN = "effect_spawn"
  final val MSG_EFFECT_UPDATE = "effect_update"

}

import cn.academy.ability.api.AbilityAPIExt._
import cn.lambdalib2.util.RandUtils._
import scala.collection.JavaConversions._
import EMContext._

class EMContext(p: EntityPlayer) extends Context(p, ElectronMissile) {

  private val MAX_HOLD: Int = 5
  
  private var active: java.util.LinkedList[EntityMdBall] = _
  private var ticks: Int = 0

  private val exp: Float = ctx.getSkillExp

  private val consumption: Float = lerpf(12, 5, exp)
  private val overload_attacked: Float = lerpf(9, 4, exp)
  private val consumption_attacked: Float = lerpf(60, 25, exp)
  private val overload_keep = 200

  private var overloadKeep = 0f

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  private def l_onEnd() = {
    terminate()
  }

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  private def l_onAbort() = {
    terminate()
  }

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.SERVER))
  private def s_madeAlive() = {
    ctx.consume(overload_keep, 0)
    overloadKeep = ctx.cpData.getOverload
    active = new java.util.LinkedList[EntityMdBall]()
  }

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  private def s_onTick() = {
    if(ctx.cpData.getOverload < overloadKeep) ctx.cpData.setOverload(overloadKeep)
    if (!ctx.consume(0, consumption)) terminate()
    else {
      val timeLimit: Int = lerpf(80, 200, exp).toInt
      if (ticks <= timeLimit) {
        if (ticks % 10 == 0) if (active.size < MAX_HOLD) {
          val ball: EntityMdBall = new EntityMdBall(player)
          player.world.spawnEntity(ball)
          active.add(ball)
        }
        if (ticks != 0 && ticks % 8 == 0) {
          val range: Float = lerpf(5, 13, exp)
          val list: java.util.List[Entity] = WorldUtils.getEntities(
            player, range,
            EntitySelectors.exclude(player).and(EntitySelectors.living)
          )
          if (!active.isEmpty && !list.isEmpty && ctx.consume(overload_attacked, consumption_attacked)) {
            var min: Double = Double.MaxValue
            var result: Entity = null
            import scala.collection.JavaConversions._
            for (e <- list) {
              val dist: Double = e.getDistanceSq(player)
              if (dist < min) {
                min = dist
                result = e
              }
            }
            // Find a random ball and destroy it
            var index: Int = 1 + nextInt(active.size)
            val iter: java.util.Iterator[EntityMdBall] = active.iterator
            var ball: EntityMdBall = null
            while ( {
              index -= 1; index + 1
            } > 0) ball = iter.next
            iter.remove()
            // client action
            sendToClient(MSG_EFFECT_SPAWN, VecUtils.entityPos(ball), VecUtils.add(VecUtils.entityPos(result),
              new Vec3d(0, result.getEyeHeight, 0)))
            // server action
            result.hurtResistantTime = -1
            val damage: Float = lerpf(10, 18, exp)
            MDDamageHelper.attack(ctx, result, damage)
            ctx.addSkillExp(0.001f)
            ball.setDead()
          }
        }
      } else {
        // ticks > timeLimit
        terminate()
      }
      sendToClient(MSG_EFFECT_UPDATE)
      ticks += 1
    }
  }

  @Listener(channel=MSG_TERMINATED, side=Array(Side.SERVER))
  private def s_onEnd() = {
    val cooldown: Int = MathUtils.clampi(700, 400, exp.toInt)
    ctx.setCooldown(cooldown)

    for (ball <- active) {
      ball.setDead()
    }
  }
  
}

@SideOnly(Side.CLIENT)
@RegClientContext(classOf[EMContext])
class EMContextC(par: EMContext) extends ClientContext(par) {

  @Listener(channel=MSG_EFFECT_UPDATE, side=Array(Side.CLIENT))
  private def c_updateEffect() = {
    var count: Int = rangei(1, 3)
    while ( {
      count -= 1; count + 1
    } > 0) {
      val r: Double = ranged(0.5, 1)
      val theta: Double = ranged(0, Math.PI * 2)
      val h: Double = ranged(-1.2, 0)
      val pos: Vec3d = VecUtils.add(new Vec3d(player.posX, player.posY + ACRenderingHelper.getHeightFix(player),
        player.posZ), new Vec3d(r * Math.sin(theta), h, r * Math.cos(theta)))
      val vel: Vec3d = new Vec3d(ranged(-.02, .02), ranged(.01, .05), ranged(-.02, .02))
      player.world.spawnEntity(MdParticleFactory.INSTANCE.next(player.world, pos, vel))
    }
  }

  @Listener(channel=MSG_EFFECT_SPAWN, side=Array(Side.CLIENT))
  private def c_spawnRay(from: Vec3d, to: Vec3d) = {
    val ray: EntityMdRaySmall = new EntityMdRaySmall(world)
    ray.setFromTo(from, to)
    world.spawnEntity(ray)
  }

}